### PR TITLE
Add forwarders for 'name()' method for Funcs, ImageParam

### PR DESCRIPTION
**Pull request by steven-johnson**
(followup to #6772)

### CODE REVIEW

File: src/Generator.h [hunk 1]
```diff
@@ -2334,9 +2337,6 @@ class GeneratorOutputBase : public GIOBase {
 
     // }@
 
-#undef HALIDE_OUTPUT_FORWARD
```
Comment (abadams) [inline src/Generator.h:2337]: Why this change? Is the macro used later?

Comment (steven-johnson) [reply src/Generator.h:2337]: Oh, sorry, drive-by change -- apparently a holdover from way back when there were multiple forwarding macros, just dead code now

**Comment by steven-johnson** [comment]
OK, I obviously didn't even test this locally because I thought it was a no-brainer, silly me

**Comment by steven-johnson** [comment]
Had to rename the inherited `GIOBase::name()` method (as I did for `type()` yesterday) to avoid ambiguity with the new forwards, PTAL

**Comment by abadams** [comment]
Hrm, I guess the name of an Output<Buffer> doesn't have to match the name of the Func you assign to it. As things currently stand, does calling name() return the Output name (i.e. gio_name)? Does this change it to return the Func name?

**Comment by abadams** [comment]
That seems surprising. My expectation would have been that if you declare an Output<Buffer> with a name, calling name() on it gives you that name back, and you have to do something like my_output_buffer.func().name() or Func(my_output_buffer).name() to get the name of a Func.

**Comment by steven-johnson** [comment]
> My expectation would have been that if you declare an Output with a name, calling name() on it gives you that name back

OK, but keep in mind that the (current) Python generator design has no special `name()` method on inputs or outputs -- the only such method available is the one on the underlying Func, ImageParam, etc. -- so this change brings the C++ behavior closer to the (proposed) Python design.

Also keep in mind that even in C++, the `gio_name` of all Inputs and Outputs is 100% fixed at compiletime; there is no way to replace those names via GeneratorParam or whatnot, so having a `name()` getter is something that would rarely be used.

**Comment by steven-johnson** [comment]
Monday Morning Review Ping

**Comment by steven-johnson** [comment]
I'm just going to drop this PR entirely: the only usage of the GIOBase::name() function that I'm aware of was in one piece of Google code, and we changed it to use a hardcoded name instead (because that's essentially what it was using). I presume that if/when other folks realize that the `name()` method is no longer public, they will let us know whether it's a problem for them.

**Pull request closed without merging**